File: Python3/1 - 200/22 Generate Parentheses.py
# 44ms, 100.0%
class Solution:
    def generateParenthesis(self, n):
        def add_parentheses(n, m):    # same runtime when nested class is not used
            if not n and not m: res.append(''.join(s))
            s.append('(')
            if n: add_parentheses(n - 1, m + 1)
            s.pop()
            s.append(')')
            if m: add_parentheses(n, m - 1)
            s.pop()
        
        res, s = [], []
        add_parentheses(n, 0)
        return res
    

File: Cpp/1 - 200/22 Generate Parentheses.cpp
// 3ms, 9.67%
class Solution
{
private:
    vector<string> ans;
    string str = "";
    
public:
    vector<string> generateParenthesis(int n)
    {
        addParenthesis(n, 0);
        return ans;
    } 
    
    void addParenthesis(int n, int m)
    {
        if(!n && !m) ans.push_back(str);
        str.push_back('(');
        if(n > 0) addParenthesis(n - 1, m + 1);
        str.pop_back();
        str.push_back(')');
        if(m > 0) addParenthesis(n, m - 1);
        str.pop_back();
    }
};
